etw/rust/tracelogging/src/lib.rs

```diff
@@ -0,0 +1,869 @@
+///     str8("MyField2", message, format(Json)), // Using the format option
+///     str8("MyField3", message, tag(0x1234)),  // Using the tag option
+///     str8("MyField4", message, format(Json), tag(0x1234)), // Both options
+/// );
+/// ```
+///
+/// ### Normal field types
+///
+/// | Field Type | Rust Type | ETW Type
+/// |------------|-----------|---------
+/// | `binary` | `&[u8]` | `Binary`
+/// | `binaryc` [^binaryc] | `&[u8]` | `BinaryC`
+/// | `bool8` | `&bool` | `U8 + Boolean`
+/// | `bool8_slice` | `&[bool]` | `U8 + Boolean`
+/// | `bool32` | `&i32` | `Bool32`
+/// | `bool32_slice` | `&[i32]` | `Bool32`
+/// | `char8_cp1252` | `&u8` | `U8 + String`
+/// | `char8_cp1252_slice` | `&[u8]` | `U8 + String`
+/// | `char16` | `&u16` | `U16 + String`
+/// | `char16_slice` | `&[u16]` | `U16 + String`
+/// | `codepointer` | `&usize` | `HexSize + CodePointer`
+/// | `codepointer_slice` | `&[usize]` | `HexSize + CodePointer`
+/// | `f32` | `&f32` | `F32`
+/// | `f32_slice` | `&[f32]` | `F32`
+/// | `f64` | `&f64` | `F64`
+/// | `f64_slice` | `&[f64]` | `F64`
+/// | `guid` | `&tracelogging::Guid` | `Guid`
+/// | `guid_slice` | `&[tracelogging::Guid]` | `Guid`
+/// | `hresult` | `&i32` | `I32 + HResult`
+/// | `hresult_slice` | `&[i32]` | `I32 + HResult`
+/// | `i8` | `&i8` | `I8`
+/// | `i8_slice` | `&[i8]` | `I8`
+/// | `i8_hex` | `&i8` | `U8 + Hex`
+/// | `i8_hex_slice` | `&[i8]` | `U8 + Hex`
+/// | `i16` | `&i16` | `I16`
+/// | `i16_slice` | `&[i16]` | `I16`
+/// | `i16_hex` | `&i16` | `U16 + Hex`
+/// | `i16_hex_slice` | `&[i16]` | `U16 + Hex`
+/// | `i32` | `&i32` | `I32`
+/// | `i32_slice` | `&[i32]` | `I32`
+/// | `i32_hex` | `&i32` | `Hex32`
+/// | `i32_hex_slice` | `&[i32]` | `Hex32`
+/// | `i64` | `&i64` | `I64`
+/// | `i64_slice` | `&[i64]` | `I64`
+/// | `i64_hex` | `&i64` | `Hex64`
+/// | `i64_hex_slice` | `&[i64]` | `Hex64`
+/// | `ipv4` | `&[u8; 4]` | `U32 + IPv4`
+/// | `ipv4_slice` | `&[[u8; 4]]` | `U32 + IPv4`
+/// | `ipv6` | `&[[u8; 16]]` | `Binary + IPv6`
+/// | `ipv6c` [^binaryc] | `&[[u8; 16]]` | `BinaryC + IPv6`
+/// | `isize` | `&isize` | `ISize`
+/// | `isize_slice` | `&[isize]` | `ISize`
+/// | `isize_hex` | `&isize` | `HexSize`
+/// | `isize_hex_slice` | `&[isize]` | `HexSize`
+/// | `pid` | `&u32` | `U32 + Pid`
+/// | `pid_slice` | `&[u32]` | `U32 + Pid`
+/// | `pointer` | `&usize` | `HexSize`
+/// | `pointer_slice` | `&[usize]` | `HexSize`
+/// | `port` | `&u16` | `U16 + Port`
+/// | `port_slice` | `&[u16]` | `U16 + Port`
+/// | `socketaddress` | `&[u8]` | `Binary + SocketAddress`
+/// | `socketaddressc` [^binaryc] | `&[u8]` | `BinaryC + SocketAddress`
+/// | `str8` | `&[u8]` | `Str8 + Utf8`
+/// | `str8_cp1252` | `&[u8]` | `Str8`
+/// | `str8_json` | `&[u8]` | `Str8 + Json`
+/// | `str8_xml` | `&[u8]` | `Str8 + Xml`
+/// | `str16` | `&[u16]` | `Str16`
+/// | `str16_json` | `&[u16]` | `Str16 + Json`
+/// | `str16_xml` | `&[u16]` | `Str16 + Xml`
+/// | `strz8` [^strz] | `&[u8]` | `StrZ8 + Utf8`
```

Comment:
Does `strz` convey the correct semantics (input is a slice that may or may not be nul-terminated, logged as a nul-terminated string)? Is there a better type name? `cstr`? `zstr`?

Reply:
I think I prefer cstr. Still open to other suggestions.